codegen/smithy-aws-swift-codegen/src/main/kotlin/software/amazon/smithy/aws/swift/codegen/customization/RulesBasedAuthSchemeResolverGenerator.kt

```diff
@@ -0,0 +1,147 @@
+        val serviceSpecificAuthResolverProtocol = sdkId + AUTH_SCHEME_RESOLVER
+
+        writer.apply {
+            writer.openBlock(
+                "public struct \$L: \$L {",
+                "}",
+                serviceSpecificDefaultResolverName,
+                serviceSpecificAuthResolverProtocol
+            ) {
+                renderResolveAuthSchemeMethod(ctx, writer)
+                write("")
+                renderConstructParametersMethod(
+                    ctx,
+                    sdkId + ClientRuntimeTypes.Auth.AuthSchemeResolverParams.name,
+                    writer
+                )
+            }
+        }
+    }
+
+    private fun renderResolveAuthSchemeMethod(ctx: ProtocolGenerator.GenerationContext, writer: SwiftWriter) {
+        val sdkId = AuthSchemeResolverGenerator.getSdkId(ctx)
+        val serviceParamsName = sdkId + ClientRuntimeTypes.Auth.AuthSchemeResolverParams.name
+
+        writer.apply {
+            openBlock(
+                "public func resolveAuthScheme(params: \$L) throws -> [AuthOption] {",
+                "}",
+                ClientRuntimeTypes.Auth.AuthSchemeResolverParams
+            ) {
+                // Return value of array of auth options
+                write("var validAuthOptions = [AuthOption]()")
+
+                // Cast params to service specific params object
+                openBlock(
+                    "guard let serviceParams = params as? \$L else {",
+                    "}",
+                    serviceParamsName
+                ) {
+                    write("throw ClientError.authError(\"Service specific auth scheme parameters type must be passed to auth scheme resolver.\")")
+                }
+
+                // Construct endpoint params from auth params
+                write("let endpointParams = EndpointParams(authSchemeParams: serviceParams)")
+                // Resolve endpoint, and retrieve auth schemes valid for the resolved endpoint
+                write("let endpoint = try DefaultEndpointResolver().resolve(params: endpointParams)")
+                openBlock("guard let authSchemes = endpoint.authSchemes() else {", "}") {
+                    // Call internal modeled model-based auth scheme resolver as fall-back if no auth schemes
+                    // are returned by endpoint resolver.
+                    write("return try InternalModeled${sdkId + AUTH_SCHEME_RESOLVER}().resolveAuthScheme(params: params)")
+                }
+                writer.write("let schemes = try authSchemes.map { (input) -> AWSClientRuntime.AuthScheme in try AWSClientRuntime.AuthScheme(from: input) }")
+                // If endpoint resolver returned auth schemes, iterate over them and save each as valid auth option to return
+                openBlock("for scheme in schemes {", "}") {
+                    openBlock("switch scheme {", "}") {
+                        // SigV4 case
+                        write("case .sigV4(let param):")
```

Comment:
S3 has a special case for an auth scheme named `sigv4-s3express`, which is equivalent to `aws.auth#sigv4a`.

Reply:
Link to ticket: https://github.com/awslabs/aws-sdk-swift/issues/1365